fontir/src/feature_variations.rs

```diff
@@ -186,6 +187,70 @@ impl Region {
     }
 }
 
+/// For emulating the behaviour of a python int, which has arbitrary precision.
+///
+/// This is necessary because the python code we are emulating does some bit
+/// fiddling, and the required number of bits may be larger than a standard word.
+///
+/// This solution doesn't handle all eventualities; it can count 511 items,
+/// and so will fail if there are more than 511 unique positions assigned
+/// to bracket layers.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Default)]
+struct Rank([u128; 4]);
+
+impl Rank {
+    const ZERO: Self = Self([0, 0, 0, 0]);
+    const MAX: usize = 128 * 4 - 1;
+
+    fn new(val: usize) -> Self {
+        let val = val.min(Self::MAX);
+        let mut this = Self::default();
+        let column = val / 128;
+        let bit = val % 128;
+        this.0[3 - column] = 1 << bit;
+        this
+    }
+
+    fn count_zeros(&self) -> u32 {
+        self.0.into_iter().map(u128::count_zeros).sum()
+    }
+
+    fn first_bit_is_set(&self) -> bool {
+        (self.0[3] & 1) > 0
+    }
+
+    // we don't need the general case, so just handle this.
+    fn right_shift_one(&mut self) {
+        let mut carry_bit = 0;
+        for val in &mut self.0 {
+            let next_carry = *val & 1;
+            *val >>= 1;
+            *val |= carry_bit << 127;
+            carry_bit = next_carry;
+        }
+    }
+}
+
+impl BitOr for Rank {
+    type Output = Self;
+
+    fn bitor(self, rhs: Self) -> Self::Output {
+        let [a1, b1, c1, d1] = self.0;
+        let [a2, b2, c2, d2] = rhs.0;
+        Self([a1 | a2, b1 | b2, c1 | c2, d1 | d2])
+    }
+}
+
+impl BitOrAssign for Rank {
+    fn bitor_assign(&mut self, rhs: Self) {
+        let [a2, b2, c2, d2] = rhs.0;
+        self.0[0] |= a2;
```

Comment:
Can you use loops, and make this generic the whole implementation generic to the number of items? Sounds easier to me.